Code Refactoring

prabha-git opened the pull request:
1. Separation of code by Model Provider. This allows adding more providers without a complex if-else condition check.

3. Added simple evaluation scoring based on substring match, With the default simple question and answer I am checking for the word "sandwich' and 'dolores' in the response. Scores would be 0 or 1.  The default is still gpt4, we can use this method to reduce API cost.

gkamradt commented:
@prabha-git this is awesome - seriously so cool to see your support on this. Thank you!

Two things I wanted to check in on
1. What do you think about an warning if the user changes the default needle and selects a non-gpt4 evaluation method? Should we make it an enum (rather than saying GPT-4 or anything else)
https://github.com/gkamradt/LLMTest_NeedleInAHaystack/pull/7/commits/8938e86344e86184f24db41d2a6305fd50c077c5#diff-05a09c80927a7bafbb8521c43d34eb8398759f066eb0e1a75a609873aaa82116R332

I ask because the 'sandwich' and 'dolores' are hard coded which will be awkward if they switch eval methods

2. I ran this for OpenAI and ran fine, then I tried Anthropic and got an error with a .pop() typo
https://github.com/gkamradt/LLMTest_NeedleInAHaystack/pull/7/commits/8938e86344e86184f24db41d2a6305fd50c077c5#diff-d2cfe243e36bf6df56a149432841389e7c611f2960763812f6c7b2b2ac154dffR31

I fixed that but then got another error for Anthropic 
```    super().__init__(**kwargs)
TypeError: LLMNeedleHaystackTester.__init__() got an unexpected keyword argument 'model_provider'```

Do you mind double checking the anthropic use case or adding a test we can run through quickly?

btw once this is merged I'd love to buy you coffee or lunch if you're in SF

prabha-git commented:
@gkamradt  - I would love to meet for a coffee or lunch but I am based in Dallas, Let me know if you are in town 😄 ☕ 

Regarding the code updates:

    1. I've introduced a new parameter called substr_validation_words, which defaults to ['dolores', 'sandwich']. Additionally, I've implemented a check to ensure that if any of the words in substr_validation_words are missing from the needle, an exception will be raised.

    2. I've addressed the error you pointed out in Anthropic Code. Unfortunately, I haven't been able to do a full test myself as I'm still awaiting the API key. Could you possibly run it on your end to confirm everything is functioning correctly?

I'm open to any further comments or thoughts you might have.

geronimi73 commented:
hey everyone, will this PR be merged?

justHungryMan commented:
This is an essential option.

pavelkraleu commented on AnthropicEvaluator.py:
```diff
@@ -0,0 +1,68 @@
+        return self.tokenizer.decode(encoded_context)
+
+    def get_prompt(self, context):
+        with open('Anthropic_prompt.txt', 'r') as file:
```
I noticed `OpenAIEvaluator` has [values](https://github.com/gkamradt/LLMTest_NeedleInAHaystack/pull/7/files#diff-11ecb358dc120bc9366ed9910773344e8249df3f4b3f45f3c0c1f279d538ce1bR51) like this hardcoded. To maintain consistency you may consider also having those values in `AnthropicEvaluator` and removing `Anthropic_prompt.txt`

prabha-git replied:
Good point, I have updated the code to add the prompt directly under `get_prompt` method. 

[Text Completion API](https://docs.anthropic.com/claude/reference/complete_post)  in Anthropic is deprecated, their recommendation is to use Message API. I have updated the code to use Message API following this [migration guide](https://docs.anthropic.com/claude/reference/migrating-from-text-completions-to-messages)

pavelkraleu commented on LLMNeedleHaystackTester.py:
```diff
@@ -166,28 +146,13 @@ async def run_test(self):
-                }
-            ]
+    @abstractmethod
+    def get_prompt(self, context):
```
Perhaps you could consider merging `get_prompt()` and `get_response_from_model()` into a single function? 

I understand the [need for precise timing](https://github.com/gkamradt/LLMTest_NeedleInAHaystack/pull/7/files#diff-05a09c80927a7bafbb8521c43d34eb8398759f066eb0e1a75a609873aaa82116R170) measurements, but I believe `get_prompt()` is so fast that it doesn't matter. 🙂

prabha-git replied:
Thank you for your suggestion regarding the merging of get_prompt() and get_response_from_model() into a single function. I appreciate the consideration given to the efficiency of the code. However, prompt engineering is a critical aspect of eliciting high-quality responses from Language Models, and the crafting of prompts can vary significantly across different models.

Maintaining a dedicated get_prompt function provides a clear and centralized location for prompt development and refinement. This separation not only ensures clarity in our codebase but also offers an organized approach for future updates and customizations specific to prompt formulation. For the time being, I believe that keeping get_prompt as an independent function is beneficial for both our current workflow and potential modifications that may arise as we continue to enhance our system's capabilities.

I hope this clarifies the rationale behind the current design choice. I'm open to further discussions on this matter as we progress and as the needs of our project evolve.

pavelkraleu commented on LLMNeedleHaystackTester.py:
```diff
@@ -200,70 +165,55 @@ async def evaluate_and_log(self, context_length, depth_percent):
-            print (f"Depth: {depth_percent}%")
-            print (f"Score: {score}")
-            print (f"Response: {response}\n")
+            print(f"-- Test Summary -- ")
```
Consider removing `f`

prabha-git replied:
fixed.

pavelkraleu commented on LLMNeedleHaystackTester.py:
```diff
@@ -1,53 +1,54 @@
-                 seconds_to_sleep_between_completions = None,
-                 print_ongoing_status = True):
-        """        
+                 substr_validation_words=['dolores', 'sandwich'],
```
Do we need all those arguments? Are they ever used by someone? 🙂

Consider moving them to a constants.

pavelkraleu commented on OpenAIEvaluator.py:
```diff
@@ -0,0 +1,71 @@
+    # Tons of defaults set, check out the LLMNeedleHaystackTester's init for more info
+    ht = OpenAIEvaluator(model_name='gpt-4-1106-preview', evaluation_method='gpt4')
+
+    ht.start_test()
```
Consider using `pre-commit` with `end-of-file-fixer` to fix warnings like this one.

prabha-git replied:
Thanks for the suggestion , added `trailing-whitespace` and `end-of-file-fixer` to pre-commit

<img width="689" alt="image" src="https://github.com/gkamradt/LLMTest_NeedleInAHaystack/assets/3776681/eb340e33-eb88-46af-972a-0e7efbbbfa15">

pavelkraleu commented on LLMNeedleHaystackTester.py:
```diff
@@ -200,70 +165,55 @@ async def evaluate_and_log(self, context_length, depth_percent):
+            results['file_name']: context_file_location
 
             # Save the context to file for retesting
             if not os.path.exists('contexts'):
```
Consider using `pathlib`, code below could be rewritten to something like

```python
contexts_dir = Path('contexts')
contexts_dir.mkdir(parents=True, exist_ok=True)

context_file_path = contexts_dir / f'{context_file_location}_context.txt'
context_file_path.write_text(context)
```

prabha-git replied:
Updated the code to use `pathlib` instead of `os.path` in the latest pull request.

pavelkraleu commented on LLMNeedleHaystackTester.py:
```diff
@@ -62,93 +63,72 @@ def __init__(self,
-        self.model_to_test_description = model_name
-        self.evaluation_model = ChatOpenAI(model="gpt-4", temperature=0, openai_api_key = self.openai_api_key)
+
+        if evaluation_method == 'gpt4':
```
I see `'gpt4'` very often, maybe we can consider having variable for example `OpenAIEvaluator.MODEL_NAME = 'gpt4'`.  
This way we may easily add support for more models like `GPT-4 Turbo` by subclassing `OpenAIEvaluator` and setting few variables.

prabha-git replied:
Thank you for the feedback on the model specification. You're correct that the "Needle in Haystack" tester is designed to be versatile and accepts a range of models as runtime parameters. Users have the flexibility to test with various models such as GPT-3.5, GPT-4, or GPT-4 Turbo, catering to diverse testing needs without any hardcoded constraints.

Regarding the model response evaluation, we currently support two methods:

substring_match: A straightforward and effective method for simpler cases.
Utilizing GPT-4: Leveraging GPT-4's advanced capabilities for more complex evaluations.
I concur with the suggestion to avoid hardcoding the model name in the code. Moving towards defining the model as a configurable constant would indeed enhance code modularity and maintainability. This change will also align with our aim for greater flexibility and adaptability in the testing framework

Created issue #10 to track this enhancement.

pavelkraleu commented on LLMNeedleHaystackTester.py:
```diff
@@ -389,14 +336,23 @@ def evaluate_response(self, response):
+        pass
+
+    @abstractmethod
+    def get_decoding(self, encoded_context):
```
I'm afraid this approach may not work for Gemini models, because we do not have public tokenizer for them.

prabha-git replied:
Although Gemini's tokenizer isn't public, our use case for tokenization—to identify fact placement depth—is effectively served by public tokenizers like 'tiktoken'. This ensures our test integrity without dependency on google's tokenizer.

pavelkraleu commented on AnthropicEvaluator.py:
```diff
@@ -0,0 +1,68 @@
+
+        if 'model_name' not in kwargs:
+            raise ValueError("model_name must be supplied with init")
+        elif "claude" not in kwargs['model_name']:
```
Consider using `"` or `'`. We may also consider using [Black](https://github.com/psf/black) to fix those inconsistencies automatically.

prabha-git replied:
fixed the inconsistencies.

<img width="657" alt="image" src="https://github.com/gkamradt/LLMTest_NeedleInAHaystack/assets/3776681/081452a1-4f28-47a1-add9-7af4b381b19f">

pavelkraleu reviewed (commented):
I think your changes are very good and will move this project forward.  
Consider my comments as ideas for future improvements 🙂

prabha-git commented:
> I think your changes are very good and will move this project forward. Consider my comments as ideas for future improvements 🙂

Thank you for your valuable feedback and for considering my changes positively! I'm eager to implement these ideas. Would you prefer I incorporate these improvements into the current pull request, or should I address them in a subsequent one?
@pavelkraleu

pavelkraleu commented:
> > I think your changes are very good and will move this project forward. Consider my comments as ideas for future improvements 🙂
> 
> Thank you for your valuable feedback and for considering my changes positively! I'm eager to implement these ideas. Would you prefer I incorporate these improvements into the current pull request, or should I address them in a subsequent one? @pavelkraleu

Its up to @gkamradt to decide, I'm nobody here 🙂

Personally, I would:

1. Merge the PR as it is, assuming it does not break anything (I have not run the code) 
3. Create Issues for my comments, if you agree with them
4. Go through Issues one by one

gkamradt commented:
Hey crew! just wanted to give an update from my end - I'm so stoked about the energy around this PR.

I recently posted a [tweet](https://twitter.com/GregKamradt/status/1762178200569057308) for a co-maintainer on this project so we can take it to the next level.

Expect a note from me within the next week on next steps.

Progress will be made!

pavelkraleu commented:
Hi @prabha-git, we had an internal discussion regarding #8 and this PR. Both PRs were taking conflicting approaches to solve the issue of separation of model interactions.

While you had chosen inheritance, #8 opted for composition, which we think is better suited for future project needs.

Since you have done a lot of work on this PR, we would appreciate it if you could address some of the issues that are essentially tasks you have already completed in this PR:

* [Install pre-commit with end-of-file-fixer](https://github.com/gkamradt/LLMTest_NeedleInAHaystack/issues/11)
* [Replace os.path with Pathlib](https://github.com/gkamradt/LLMTest_NeedleInAHaystack/issues/12)
* [Update package Anthropic](https://github.com/gkamradt/LLMTest_NeedleInAHaystack/issues/13)

Sorry for the confusion. We are looking forward to your contributions to the issues I mentioned above.

prabha-git commented:
Ok, np, that Sounds good. I will start with #13 .  Lets close this pull request.

pavelkraleu commented:
Thank you for looking into that Issue.
Sorry about those complications. I appreciate your efforts in this PR.
